rfcs/20180824-tf-print-v2.md

```diff
@@ -0,0 +1,549 @@
+there is need to wrap parts of execution and optimized graph functions
+(currently `tf.contrib.eager.defun`). The Python `print` method placed in these
+graph function will run at graph construction time, not graph execution time,
+creating confusion.
+
+This doc proposes a new `tf.print` TensorFlow printing approach that is very
+similar to the standard python `print` API whether or not executing eagerly. It
+also provides long-requested functionality for both eager and session-based
+execution, such as a more meaningful Tensor summarization, support for printing
+nested data structures that contain tensors, and controllable logging levels.
+
+## Overview
+
+We introduce two methods to the public Python API: `tf.strings.format` and
+`tf.print`. (We may have to use `tf.strings.Format` or `tf.strings.fmt` instead
+of `tf.strings.format` if the built-in python `format` method causes issues with
+the naming.) These are internally backed by two C++ operators: `StringFormat`
+and `PrintV2`. The method headers & docstrings are as follows:
+
+### Python Public API Methods
+
+````python
+@tf_export("print")
+def print_v2(*inputs, **kwargs):
+  """Print the specified inputs.
+  Returns an operator that prints the specified inputs to a desired
+  output stream or logging level. The inputs may be dense or sparse Tensors,
+  primitive python objects, data structures that contain Tensors, and printable
+  python objects. Printed tensors will recursively show the first and last
+  `summarize` elements of each dimension.
+
+  With eager execution enabled and/or inside a `tf.contrib.eager.defun` this
+  operator will automatically execute, and users only need to call `tf.print`
+  without using the return value. When constructing graphs outside of a
+  `tf.contrib.eager.defun`, one must either include the returned op
+  in the input to `session.run`, or use the operator as a control dependency for
+  executed ops by specifying `with tf.control_dependencies([print_op])`.
+
+  @compatibility(python2)
+  In python 2.7, make sure to import the following:
+  `from __future__ import print_function`
+  @end_compatibility
+
+  Example:
+    Single-input usage:
+    ```python
+    tf.enable_eager_execution()
+    tensor = tf.range(10)
+    tf.print(tensor, output_stream=sys.stderr)
+    ```
+    (This prints "[0 1 2 ... 7 8 9]" to sys.stderr)
+
+    Multi-input usage:
+    ```python
+    tf.enable_eager_execution()
+    tensor = tf.range(10)
+    tf.print("tensors:", tensor, {2: tensor * 2}, output_stream=sys.stdout)
+    ```
+    (This prints "tensors: [0 1 2 ... 7 8 9] {2: [0 2 4 ... 14 16 18]}" to sys.stdout)
+
+    Usage when constructing graphs:
+    ```python
+    sess = tf.Session()
+    with sess.as_default():
+        tensor = tf.range(10)
+        print_op = tf.print("tensors:", tensor, {2: tensor * 2},
+                            output_stream=sys.stdout)
+        with tf.control_dependencies([print_op]):
```

Comment:
From a session-based graph building perspective those problems do continue to exist. My understanding is this is a lower priority to solve if we're moving to an eager & graph function-first world.

We could also add support for `tf.print` as a context manager in session-based graph building mode, so users could say:
```
with tf.print(...):
    ...
    out = ...
session.run(out)
```
Instead of having to explicitly add it with tf.control_dependencies.

Alternatively, it might be better to suggest that users add the return value of tf.print to their `session.run` code rather than add control_dependencies, if that's a better user experience. That would be harder w/ control flow and functions in graph mode though.

We could even try implicitly adding `tf.print`s to a graph collection that automatically gets executed at session.run. I'm more hesitant to do that, because it would require a lot of care with control flow & functions, and to work nicely with graph pruning.

Reply:
FWIW, I agree with @karmel and feel that the new proposal is a worse experience for session-based graph users. If solving this problem is a low priority for the team, please reconsider deprecating `tf.Print` until a better solution is available.